File: src/main/java/com/abysalto/middleware/source/dummyjson/DummyJsonProductSource.java
package com.abysalto.middleware.source.dummyjson;

import com.abysalto.middleware.domain.Product;
import com.abysalto.middleware.domain.ProductPage;
import com.abysalto.middleware.exception.ProductNotFoundException;
import com.abysalto.middleware.exception.UpstreamException;
import com.abysalto.middleware.source.ProductSource;
import com.abysalto.middleware.source.dummyjson.dto.DummyProduct;
import com.abysalto.middleware.source.dummyjson.dto.DummyProductList;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.client.ClientHttpResponse;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestClient;
import org.springframework.web.client.RestClientResponseException;

import java.io.IOException;
import java.util.List;
import java.util.function.Supplier;

/**
 * {@link ProductSource} backed by the DummyJSON REST API.
 *
 * <p>Filter/search push-down policy (documented in the README):
 * <ul>
 *   <li>Name search and category filtering are pushed down to the upstream
 *       ({@code /products/search}, {@code /products/category/{slug}}).</li>
 *   <li>Pagination uses the upstream {@code limit}/{@code skip} parameters.</li>
 *   <li>Price-range filtering is <em>not</em> supported by DummyJSON, so it is applied
 *       in-service (see {@code ProductService}); this source only exposes the primitives.</li>
 * </ul>
 *
 * <p>Calls are synchronous: the middleware runs on the servlet (Web MVC) stack and uses the blocking
 * {@link RestClient}, so no reactive types leak beyond this class.
 */
@Component
public class DummyJsonProductSource implements ProductSource {

    private static final Logger log = LoggerFactory.getLogger(DummyJsonProductSource.class);

    private final RestClient restClient;

    public DummyJsonProductSource(RestClient dummyJsonRestClient) {
        this.restClient = dummyJsonRestClient;
    }

    @Override
    public ProductPage list(int skip, int limit) {
        log.debug("Upstream list: skip={}, limit={}", skip, limit);
        DummyProductList body = fetch("list products", () -> restClient.get()
                .uri(uri -> uri.path("/products")
                        .queryParam("limit", limit)
                        .queryParam("skip", skip)
                        .build())
                .retrieve()
                .onStatus(HttpStatusCode::isError, (request, response) -> {
                    throw upstreamError(response);
                })
                .body(DummyProductList.class));
        return toPage(body);
    }

    @Override
    public Product getById(long id) {
        log.debug("Upstream getById: id={}", id);
        DummyProduct body = fetch("get product " + id, () -> restClient.get()
                .uri("/products/{id}", id)
                .retrieve()
                .onStatus(status -> status.value() == 404, (request, response) -> {
                    throw new ProductNotFoundException(id);
                })
                .onStatus(HttpStatusCode::isError, (request, response) -> {
                    throw upstreamError(response);
                })
                .body(DummyProduct.class));
        return DummyProductMapper.toDomain(body);
    }

    @Override
    public ProductPage findByCategory(String category, int skip, int limit) {
        log.debug("Upstream findByCategory: category={}, skip={}, limit={}", category, skip, limit);
        DummyProductList body = fetch("filter by category " + category, () -> restClient.get()
                .uri(uri -> uri.path("/products/category/{category}")
                        .queryParam("limit", limit)
                        .queryParam("skip", skip)
                        .build(category))
                .retrieve()
                .onStatus(HttpStatusCode::isError, (request, response) -> {
                    throw upstreamError(response);
                })
                .body(DummyProductList.class));
        return toPage(body);
    }

    @Override
    public ProductPage searchByName(String query, int skip, int limit) {
        log.debug("Upstream searchByName: q={}, skip={}, limit={}", query, skip, limit);
        DummyProductList body = fetch("search products", () -> restClient.get()
                .uri(uri -> uri.path("/products/search")
                        .queryParam("q", query)
                        .queryParam("limit", limit)
                        .queryParam("skip", skip)
                        .build())
                .retrieve()
                .onStatus(HttpStatusCode::isError, (request, response) -> {
                    throw upstreamError(response);
                })
                .body(DummyProductList.class));
        return toPage(body);
    }

    @Override
    public List<String> categories() {
        log.debug("Upstream categories");
        String[] body = fetch("list categories", () -> restClient.get()
                .uri("/products/category-list")
                .retrieve()
                .onStatus(HttpStatusCode::isError, (request, response) -> {
                    throw upstreamError(response);
                })
                .body(String[].class));
        return body == null ? List.of() : List.of(body);
    }

    private ProductPage toPage(DummyProductList body) {
        if (body == null || body.products() == null) {
            return new ProductPage(List.of(), 0, 0, 0);
        }
        List<Product> items = body.products().stream().map(DummyProductMapper::toDomain).toList();
        return new ProductPage(items, body.total(), body.skip(), body.limit());
    }

    /** Maps any error HTTP status into an {@link UpstreamException}. */
    private UpstreamException upstreamError(ClientHttpResponse response) throws IOException {
        return new UpstreamException("DummyJSON returned " + response.getStatusCode());
    }

    /**
     * Executes an upstream call and normalizes failures: {@link ProductNotFoundException} propagates
     * as-is, everything else (timeouts, connection errors, non-2xx) becomes {@link UpstreamException}.
     */
    private <T> T fetch(String description, Supplier<T> call) {
        try {
            return call.get();
        } catch (ProductNotFoundException | UpstreamException e) {
            throw e;
        } catch (RestClientResponseException e) {
            log.error("Upstream error during {}: {}", description, e.getStatusCode());
            throw new UpstreamException("Upstream error during " + description + ": " + e.getStatusCode(), e);
        } catch (Exception e) {
            log.error("Upstream call failed during {}: {}", description, e.getMessage());
            throw new UpstreamException("Upstream call failed during " + description, e);
        }
    }
}

File: dotnet/src/Middleware.Infrastructure/Upstream/DummyJsonProductSource.cs
using System.Net;
using System.Net.Http.Json;
using Microsoft.Extensions.Logging;
using Middleware.Core.Abstractions;
using Middleware.Core.Domain;
using Middleware.Core.Exceptions;
using Middleware.Infrastructure.Upstream.Dto;

namespace Middleware.Infrastructure.Upstream;

/// <summary>
/// <see cref="IProductSource"/> backed by the DummyJSON REST API.
///
/// <para>Filter/search push-down policy (documented in the README):</para>
/// <list type="bullet">
///   <item>Name search and category filtering are pushed down to the upstream
///   (<c>/products/search</c>, <c>/products/category/{slug}</c>).</item>
///   <item>Pagination uses the upstream <c>limit</c>/<c>skip</c> parameters.</item>
///   <item>Price-range filtering is <em>not</em> supported by DummyJSON, so it is applied in-service
///   (see <c>ProductService</c>); this source only exposes the primitives.</item>
/// </list>
///
/// <para>Wired as a typed <see cref="HttpClient"/> whose base address and connect/response timeouts
/// come from <c>UpstreamOptions</c>, so a slow or unreachable upstream fails fast (surfacing as a 502)
/// rather than hanging the request.</para>
/// </summary>
public sealed class DummyJsonProductSource(HttpClient httpClient, ILogger<DummyJsonProductSource> logger)
    : IProductSource
{
    private static readonly ProductPage EmptyPage = new([], 0, 0, 0);

    public async Task<ProductPage> ListAsync(int skip, int limit, CancellationToken ct = default)
    {
        logger.LogDebug("Upstream list: skip={Skip}, limit={Limit}", skip, limit);
        var body = await FetchAsync<DummyProductList>(
            "list products", $"products?limit={limit}&skip={skip}", notFoundId: null, ct);
        return ToPage(body);
    }

    public async Task<Product> GetByIdAsync(long id, CancellationToken ct = default)
    {
        logger.LogDebug("Upstream getById: id={Id}", id);
        var body = await FetchAsync<DummyProduct>(
            $"get product {id}", $"products/{id}", notFoundId: id, ct);
        if (body is null)
        {
            throw new UpstreamException($"Upstream returned an empty body for product {id}");
        }
        return DummyProductMapper.ToDomain(body);
    }

    public async Task<ProductPage> FindByCategoryAsync(string category, int skip, int limit, CancellationToken ct = default)
    {
        logger.LogDebug("Upstream findByCategory: category={Category}, skip={Skip}, limit={Limit}", category, skip, limit);
        var body = await FetchAsync<DummyProductList>(
            $"filter by category {category}",
            $"products/category/{Uri.EscapeDataString(category)}?limit={limit}&skip={skip}",
            notFoundId: null, ct);
        return ToPage(body);
    }

    public async Task<ProductPage> SearchByNameAsync(string query, int skip, int limit, CancellationToken ct = default)
    {
        logger.LogDebug("Upstream searchByName: q={Query}, skip={Skip}, limit={Limit}", query, skip, limit);
        var body = await FetchAsync<DummyProductList>(
            "search products",
            $"products/search?q={Uri.EscapeDataString(query)}&limit={limit}&skip={skip}",
            notFoundId: null, ct);
        return ToPage(body);
    }

    public async Task<IReadOnlyList<string>> CategoriesAsync(CancellationToken ct = default)
    {
        logger.LogDebug("Upstream categories");
        var body = await FetchAsync<string[]>(
            "list categories", "products/category-list", notFoundId: null, ct);
        return body ?? [];
    }

    private static ProductPage ToPage(DummyProductList? body)
    {
        if (body?.Products is null)
        {
            return EmptyPage;
        }
        // Frozen, not just projected: the page is cached and shared across concurrent requests
        // (see the immutability contract on ProductPage).
        var items = DummyProductMapper.Freeze(body.Products.Select(DummyProductMapper.ToDomain))!;
        return new ProductPage(items, body.Total, body.Skip, body.Limit);
    }

    /// <summary>
    /// Executes an upstream GET and normalizes failures: a 404 with a <paramref name="notFoundId"/>
    /// becomes <see cref="ProductNotFoundException"/>, any other non-success status or transport
    /// failure (timeout, connection error) becomes <see cref="UpstreamException"/>. The two typed
    /// exceptions propagate as-is if thrown deeper.
    /// </summary>
    private async Task<T?> FetchAsync<T>(string description, string uri, long? notFoundId, CancellationToken ct)
    {
        try
        {
            using var response = await httpClient.GetAsync(uri, ct);

            if (notFoundId is not null && response.StatusCode == HttpStatusCode.NotFound)
            {
                throw new ProductNotFoundException(notFoundId.Value);
            }
            if (!response.IsSuccessStatusCode)
            {
                logger.LogError("Upstream error during {Description}: {Status}", description, (int)response.StatusCode);
                throw new UpstreamException($"DummyJSON returned {(int)response.StatusCode} {response.StatusCode}");
            }

            return await response.Content.ReadFromJsonAsync<T>(ct);
        }
        catch (ProductNotFoundException)
        {
            throw;
        }
        catch (UpstreamException)
        {
            throw;
        }
        catch (Exception e)
        {
            logger.LogError(e, "Upstream call failed during {Description}: {Message}", description, e.Message);
            throw new UpstreamException($"Upstream call failed during {description}", e);
        }
    }
}
